extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -17,44 +17,169 @@
- * com.hazelcast.jet.avrp} package. This is not a public-facing API.
+ * Hazelcast serializer hooks for the classes in the {@code com.hazelcast.jet.avro} package.
  */
-class AvroSerializerHooks {
+public final class AvroSerializerHooks {
+    private AvroSerializerHooks() { }
 
-    public static final class Utf8Hook implements SerializerHook<Utf8> {
+    public static <T> byte[] serialize(DatumWriter<T> datumWriter, T data) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+            datumWriter.write(data, encoder);
+            encoder.flush();
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericContainerHook implements SerializerHook<GenericContainer> {
+        private static final int CACHE_SIZE = 500;
+        private static final int CACHE_THRESHOLD = CACHE_SIZE + Math.min(CACHE_SIZE / 10, 50);
+
+        @Override
+        public Class<GenericContainer> getSerializationType() {
+            return GenericContainer.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new StreamSerializer<GenericContainer>() {
+                private final ReadOptimizedLruCache<String, Schema> jsonToSchema =
+                        new ReadOptimizedLruCache<>(CACHE_SIZE, CACHE_THRESHOLD);
+                private final ReadOptimizedLruCache<Schema, String> schemaToJson =
+                        new ReadOptimizedLruCache<>(CACHE_SIZE, CACHE_THRESHOLD);
+
+                @Override
+                public int getTypeId() {
+                    return SerializerHookConstants.AVRO_GENERIC_CONTAINER;
+                }
 
+                @Override
+                public void write(@Nonnull ObjectDataOutput out, @Nonnull GenericContainer datum) throws IOException {
+                    if (datum instanceof SpecificRecord) {
+                        // SpecificRecord schemas are indistinguishable from others. We cannot add a temporary
+                        // flag to the schema since schema properties are add-only. As a solution, we only write
+                        // the name of the datum class and retrieve the schema from the class. The reader can
+                        // distinguish class names from actual schemas by checking the first character: RECORD,
+                        // ARRAY, ENUM and FIXED schemas start with '{', which is invalid for class names.
+                        out.writeString(datum.getClass().getName());
```

Comment:
I can see a few problems with this approach:
1. This is yet another format of serialized data (classname or json + binary data), which we nest in IDS. We already have too many serialization formats.
2. `SpecificRecordBase` is `Externalizable` so it was already serializable. If users have any such data in IMaps this will cause problems during RU and maybe even later because you changed how `SpecificRecordBase` is serialized.
3. The implementation (apart from class name) is similar to `SpecificRecordBase.writeExternal` but I'm not sure if completely equivalent.
4. given type id (`getTypeId()`) usually matches single, specific class. I am not sure if that is required, but definitely usage here is highly unusual (AFAIK we had slightly similar discussion for supporting inheritance in IDS).

Reply:
Do we really want to serialize all implementations of `GenericContainer`. Maybe it would be sufficient to have dedicated serializers for: `GenericArray`, `GenericFixed` and `GenericRecord` (each with it's own F_ID).
For `SpecificRecord` we might assume that all generated classes will inherit from `SpecificRecordBase` which is `Externalizable` and we already have serialization support for that. If `SpecificRecord` does not implement `Externalizable`, `Serializable` etc. it will be non-serializable - that should be fine.
There are a few other branches in `GenericContainer` hierarchy, I am not sure if they are needed: GenericEnumSymbol, Pair (in mapred - likely irrelevant for us).

Reply:
`SpecificRecordBase` is also `GenericRecord`, so even `GenericRecord` is not specific enough. I think the problem  here is that, as you pointed out in 4., a serializer should be responsible for a specific class, rather than an interface. Otherwise, it may collide with other serializers, which might be added by us or users. Thus, we should only provide serializers for `GenericData$Record`, `GenericData$Array`, `GenericData$Fixed` and `GenericData$EnumSymbol`. Luckily, they are not marked as final, so lazy deserialization is still possible by subclassing.

We might actually consider relaxing the type boundaries of the deserialized object in serializers to allow returning wrappers. Or, we might add filtering feature to serializers to reject some types.

Reply:
> SpecificRecordBase is also GenericRecord

meh...

Looking at `lookupCustomSerializer`, `registerFromSuperType` and `safeRegister` I think it might be possible to register asymmetric serializer configuration. We could register `ExternalizableSerializer` for `SpecificRecordBase` (for backward compatibility) and the custom `StreamSerializer<GenericContainer>` for `GenericContainer`. According to supertype traversal logic in `lookupCustomSerializer` during serialization it should choose ExternalizableSerializer for SpecificRecordBase and `StreamSerializer<GenericContainer>` for everything else (lookupCustomSerializer considers base classes before implemented interfaces). Deserialization matches only (?) by typeId, so it should choose a correct one.

This approach might be not that bad as `registerFromSuperType` already is used to generate such assymetry: there can be less entries in `typeMap` than in `idMap`

Reply:
> We might actually consider relaxing the type boundaries of the deserialized object in serializers to allow returning wrappers.

generic type in `StreamSerializer` seems to not be so important, it is handled via `Object` anyway, so we only need to make runtime type checks satisied.

Reply:
083b8e4d